### PR TITLE
fix: allow overrides with version specs

### PR DESCRIPTION
**Pull request by JamesPatrickGill**
### What this does

Please look at test fixtures to understand goal.

This allows us to work with simple overrides which have a version specifier in the name.

### CODE REVIEW

File: lib/dep-graph-builders/npm-lock-v2/index.ts [hunk 1]
```diff
@@ -444,3 +443,42 @@ const checkOverrides = (
   }
   return;
 };
+
+// Here we have to match our pkg to
+// possible keys in the overrides object
+const matchOverrideKey = (
```
Comment (anthogez) [inline lib/dep-graph-builders/npm-lock-v2/index.ts:449]: tests for matchOverrideKey?

File: lib/dep-graph-builders/npm-lock-v2/index.ts [hunk 2]
```diff
@@ -444,3 +443,42 @@ const checkOverrides = (
...
+  overrides: Overrides,
+  pkg: { name: string; version: string },
+): string | null => {
+  if (overrides[pkg.name]) {
+    return overrides[pkg.name];
+  }
+
+  const overrideKeysNameToVersions = Object.keys(overrides).reduce(
+    (acc, key) => {
+      const splitNameAndVersion = key.split('@');
```
Comment (anthogez) [inline lib/dep-graph-builders/npm-lock-v2/index.ts:459]: will always be the case where split per `@` will be valid?

File: lib/dep-graph-builders/npm-lock-v2/index.ts [hunk 3]
```diff
@@ -444,3 +443,42 @@ const checkOverrides = (
...
   return;
 };
+
+// Here we have to match our pkg to
+// possible keys in the overrides object
+const matchOverrideKey = (
+  overrides: Overrides,
+  pkg: { name: string; version: string },
+): string | null => {
+  if (overrides[pkg.name]) {
```
[the remark this answers is not in the document]
Comment (JamesPatrickGill) [reply lib/dep-graph-builders/npm-lock-v2/index.ts:453]: If theres one with just name it wins

File: lib/dep-graph-builders/npm-lock-v2/index.ts [hunk 4]
```diff
@@ -444,3 +443,42 @@ const checkOverrides = (
...
+  const overrideKeysNameToVersions = Object.keys(overrides).reduce(
+    (acc, key) => {
+      const splitNameAndVersion = key.split('@');
+      const versionSpec = splitNameAndVersion.pop();
+      const name = splitNameAndVersion.join('@');
+      const versionsArr = acc[name]
+        ? [...acc[name], versionSpec]
+        : [versionSpec];
+      return { ...acc, [name]: versionsArr };
+    },
```
Comment (anthogez) [inline lib/dep-graph-builders/npm-lock-v2/index.ts:466]: How about?

```
  // Split the key to separate the package name from the version spec
  const atIndex = key.lastIndexOf('@');
  const name = key.substring(0, atIndex);
  const versionSpec = key.substring(atIndex + 1);

  // Check if the package name already exists in the accumulator
  if (!acc[name]) {
    acc[name] = [];
  }

  // Add the version spec to the list of versions for this package name
  acc[name].push(versionSpec);

  return acc;
```

File: lib/dep-graph-builders/npm-lock-v2/index.ts (hunk 1 above)
Comment (JamesPatrickGill) [reply lib/dep-graph-builders/npm-lock-v2/index.ts:449]: Can write a specific unit test

File: lib/dep-graph-builders/npm-lock-v2/index.ts [hunk 5]
```diff
@@ -444,3 +443,42 @@ const checkOverrides = (
...
+    {},
+  );
+
+  if (overrideKeysNameToVersions[pkg.name]) {
+    for (const versionSpec of overrideKeysNameToVersions[pkg.name]) {
```
Comment (anthogez) [inline lib/dep-graph-builders/npm-lock-v2/index.ts:471]: why are we increasing computation time twice?

```
const keysNameToVersion = overrideKeysNameToVersions[pkg.name];

then reuse keysNameToVersion both for line 470 and line 471?
```

Comment (JamesPatrickGill) [reply lib/dep-graph-builders/npm-lock-v2/index.ts:471]: Fair enough, was an extra hash lookup so didn't delve in too much

**Review by anthogez — APPROVED** [review]

**Pull request merged**